File: api/internal/repository/manual_targets_test.go
package repository_test

import (
	"testing"

	"github.com/Kenyan822/physique-analytics/api/gen/openapi"
	"github.com/Kenyan822/physique-analytics/api/internal/repository"
	"github.com/Kenyan822/physique-analytics/api/internal/testdb"
)

func TestManualTargets_設定していなければnil(t *testing.T) {
	t.Parallel()
	ctx := t.Context()
	repo := repository.NewManualTargets(testdb.Begin(t))

	got, err := repo.Get(ctx)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got != nil {
		t.Errorf("Get() = %+v, want nil", got)
	}
}

func TestManualTargets_保存して読める(t *testing.T) {
	t.Parallel()
	ctx := t.Context()
	repo := repository.NewManualTargets(testdb.Begin(t))

	in := openapi.ManualTargets{ProteinG: 180, FatG: 70, CarbG: 250}
	saved, err := repo.Put(ctx, in)
	if err != nil {
		t.Fatalf("Put: %v", err)
	}
	// kcal は列に持たず、読むときに計算する（Atwater 4/9/4）
	if saved.Kcal == nil || *saved.Kcal != 180*4+70*9+250*4 {
		t.Errorf("Kcal = %v, want %d", saved.Kcal, 180*4+70*9+250*4)
	}

	got, err := repo.Get(ctx)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil || got.ProteinG != 180 || got.CarbG != 250 {
		t.Errorf("Get() = %+v", got)
	}
}

func TestManualTargets_上書きできる(t *testing.T) {
	t.Parallel()
	ctx := t.Context()
	repo := repository.NewManualTargets(testdb.Begin(t))

	if _, err := repo.Put(ctx, openapi.ManualTargets{ProteinG: 180, FatG: 70, CarbG: 250}); err != nil {
		t.Fatalf("Put: %v", err)
	}
	// **単一行。** 2行目を作らず置き換える
	if _, err := repo.Put(ctx, openapi.ManualTargets{ProteinG: 200, FatG: 60, CarbG: 200}); err != nil {
		t.Fatalf("Put 2回目: %v", err)
	}

	got, err := repo.Get(ctx)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil || got.ProteinG != 200 {
		t.Errorf("ProteinG = %v, want 200", got)
	}
}

func TestManualTargets_消すと自動計算に戻る(t *testing.T) {
	t.Parallel()
	ctx := t.Context()
	repo := repository.NewManualTargets(testdb.Begin(t))

	if _, err := repo.Put(ctx, openapi.ManualTargets{ProteinG: 180, FatG: 70, CarbG: 250}); err != nil {
		t.Fatalf("Put: %v", err)
	}
	if err := repo.Delete(ctx); err != nil {
		t.Fatalf("Delete: %v", err)
	}

	got, err := repo.Get(ctx)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got != nil {
		t.Errorf("Get() = %+v, want nil", got)
	}
}

func TestManualTargets_無くても消せる(t *testing.T) {
	t.Parallel()
	ctx := t.Context()
	repo := repository.NewManualTargets(testdb.Begin(t))

	// 設定していない状態で消してもエラーにしない（DELETE は冪等）
	if err := repo.Delete(ctx); err != nil {
		t.Errorf("Delete: %v", err)
	}
}

File: web/app/settings/manualTargets.test.ts
import { describe, expect, it } from "vitest";

import { kcalFromMacros, toManualTargets } from "./manualTargets";

describe("PFC から kcal を出す", () => {
  it("Atwater 係数 4/9/4", () => {
    // サーバの analytics.KcalFromMacros と一致させる
    expect(kcalFromMacros(30, 10, 60)).toBe(450);
    expect(kcalFromMacros(180, 70, 250)).toBe(2350);
  });

  it("端数は四捨五入", () => {
    // 120.4 + 91.8 + 241.2 = 453.4
    expect(kcalFromMacros(30.1, 10.2, 60.3)).toBe(453);
  });

  it("全部 0", () => {
    expect(kcalFromMacros(0, 0, 0)).toBe(0);
  });
});

describe("入力を目標にする", () => {
  it("3つ揃っていれば通る", () => {
    expect(toManualTargets({ proteinG: "180", fatG: "70", carbG: "250" })).toEqual({
      proteinG: 180,
      fatG: 70,
      carbG: 250,
    });
  });

  it("前後の空白を無視する", () => {
    expect(toManualTargets({ proteinG: " 180 ", fatG: "70", carbG: "250" })?.proteinG).toBe(180);
  });

  it("小数を読む", () => {
    expect(toManualTargets({ proteinG: "180.5", fatG: "70", carbG: "250" })?.proteinG).toBe(180.5);
  });

  // **3つで1組。** 1つ欠けた目標は意味を成さない（iOS と同じ規則）
  it("**1つでも空なら null**", () => {
    expect(toManualTargets({ proteinG: "180", fatG: "", carbG: "250" })).toBeNull();
    expect(toManualTargets({ proteinG: "", fatG: "", carbG: "" })).toBeNull();
  });

  it("読めない値は null", () => {
    expect(toManualTargets({ proteinG: "だいたい180", fatG: "70", carbG: "250" })).toBeNull();
  });

  it("負は受け付けない", () => {
    expect(toManualTargets({ proteinG: "-1", fatG: "70", carbG: "250" })).toBeNull();
  });

  it("上限を超えたら null", () => {
    expect(toManualTargets({ proteinG: "1001", fatG: "70", carbG: "250" })).toBeNull();
    expect(toManualTargets({ proteinG: "180", fatG: "70", carbG: "2001" })).toBeNull();
  });
});
